jms/src/main/scala/akka/stream/alpakka/jms/JmsSourceStage.scala

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.jms
+
+import java.util.concurrent.Semaphore
+import javax.jms._
+
+import akka.stream.stage.{ GraphStage, GraphStageLogic, OutHandler, StageLogging }
+import akka.stream.{ Attributes, Outlet, SourceShape }
+
+import scala.collection.mutable
+import scala.util.{ Failure, Success }
+
+final class JmsSourceStage(settings: JmsSourceSettings) extends GraphStage[SourceShape[Message]] {
+
+  private val out = Outlet[Message]("JmsSource.out")
+
+  override def shape: SourceShape[Message] = SourceShape[Message](out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with JmsConnector with StageLogging {
+
+      override private[jms] def jmsSettings = settings
+
+      private val bufferSize = settings.bufferSize
+      private val queue = mutable.Queue[Message]()
```

Comment:
just to be nice to GC (and in case the stage is still referenced from somewhere) you could clear the queue in `postStop`